numpy/core/src/umath/loops.c.src

```diff
@@ -840,6 +847,115 @@ NPY_NO_EXPORT NPY_GCC_OPT_3 void
     UNARY_LOOP_FAST(@type@, @type@, *out = in > 0 ? 1 : (in < 0 ? -1 : 0));
 }
 
+#ifdef USE_LIBDIVIDE
+NPY_NO_EXPORT void
+@TYPE@_divide(char **args, npy_intp const *dimensions, npy_intp const *steps, void *NPY_UNUSED(func))
+{
+    BINARY_LOOP_BASE
+
+    if(!is2) {
+        const @type@ in2 = *(@type@ *)ip2;
+        struct libdivide_@div@_t fast_d = libdivide_@div@_gen(in2);
+        BINARY_LOOP_FIXED {
+            const @type@ in1 = *(@type@ *)ip1;
+            /*
+             * FIXME: On x86 at least, dividing the smallest representable integer
+             * by -1 causes a SIFGPE (division overflow). We treat this case here
+             * (to avoid a SIGFPE crash at python level), but a good solution would
+             * be to treat integer division problems separately from FPU exceptions
+             * (i.e. a different approach than npy_set_floatstatus_divbyzero()).
+             */
+            if (in2 == 0 || (in1 == NPY_MIN_@TYPE@ && in2 == -1)) {
+                npy_set_floatstatus_divbyzero();
+                *((@type@ *)op1) = 0;
+            }
+            else if (((in1 > 0) != (in2 > 0)) && (in1 % in2 != 0)) {
+                *((@type@ *)op1) = libdivide_@div@_do(in1, &fast_d) - 1;
+            }
+            else {
+                *((@type@ *)op1) = libdivide_@div@_do(in1, &fast_d);
+            }
+        }
+    }
+    else {
+        BINARY_LOOP_SLIDING {  // XXX Lot of repeated code
+            const @type@ in1 = *(@type@ *)ip1;
+            const @type@ in2 = *(@type@ *)ip2;
+            /*
+             * FIXME: On x86 at least, dividing the smallest representable integer
+             * by -1 causes a SIFGPE (division overflow). We treat this case here
+             * (to avoid a SIGFPE crash at python level), but a good solution would
+             * be to treat integer division problems separately from FPU exceptions
+             * (i.e. a different approach than npy_set_floatstatus_divbyzero()).
+             */
+            if (in2 == 0 || (in1 == NPY_MIN_@TYPE@ && in2 == -1)) {
+                npy_set_floatstatus_divbyzero();
+                *((@type@ *)op1) = 0;
+            }
+            else if (((in1 > 0) != (in2 > 0)) && (in1 % in2 != 0)) {
+                *((@type@ *)op1) = in1/in2 - 1;
+            }
+            else {
+                *((@type@ *)op1) = in1/in2;
+            }
+        }
+    }
+}
+#elif defined(USE_OPTIMAL_DIVISOR)
+NPY_NO_EXPORT void
+@TYPE@_divide(char **args, npy_intp const *dimensions, npy_intp const *steps, void *NPY_UNUSED(func))
+{
+    BINARY_LOOP_BASE
+
+    if(!is2) {
+        const @type@ in2 = *(@type@ *)ip2;
+        const float in2_f = (float) in2;
```

Comment:
Casting to float like this will definitely result in incorrect division results for integers

Reply:
#16453 gives a rough outline of the type of problem that can come up, although it's probably a little harder to construct a case for integer division

Reply:
We do floor the result as PEP suggests yeah(`a // b == floor(a/b)`), shall I look at other ways to remove the `%`

Reply:
But yeah we can remove it, seems like the `%` is not slowing things down

Reply:
Here's the failing test:
```python
>>> a = 2**28
>>> b = 2**28 - 1
>>> np.int64(a * b) / np.int64(b) == a  # Today
True
>>> np.int64(a * b) / np.float32(b) == a  # Your code
False
```

Reply:
The test is harder to construct if you'd used `float64` (double), but unless proven otherwise I'm going to assume one exists.

Reply:
Ohh nice, I was trying random smaller values. will remove it 👍